src/modules/ekf2/EKF/mag_control.cpp

```diff
@@ -159,100 +148,227 @@ void Ekf::controlMagFusion()
+				}
+			}
 		}
 
 		if (!_control_status.flags.yaw_align) {
 			// Having the yaw aligned is mandatory to continue
+			stopMagFusion();
+			stopMag3DFusion();
+			stopMagHdgFusion();
 			return;
 		}
 
+		const bool mag_3d_forced_always = (_params.mag_fusion_type == MagFuseType::MAG_3D);
+
+		if (_mag_bias_observable
+		|| _yaw_angle_observable
+		|| isRecent(_time_last_mov_3d_mag_suitable, (uint64_t)2e6)
+		|| mag_3d_forced_always
+		) {
+			startMagFusion();
+
+		} else {
+			stopMagFusion();
+		}
+
 		checkMagDeclRequired();
 
-		runMagAndMagDeclFusions(mag_sample.mag);
+
+		// mag
+		resetEstimatorAidStatus(_aid_src_mag);
+		_aid_src_mag.timestamp_sample = mag_sample.time_us;
+
+		if (_control_status.flags.mag) {
+
+			// For the first few seconds after in-flight alignment we allow the magnetic field state estimates to stabilise
+			// before they are used to constrain heading drift
+			const bool update_all_states = _control_status.flags.mag_3D
+				&& ((_control_status.flags.mag_aligned_in_flight && isTimedOut(_flt_mag_align_start_time, (uint64_t)5e6)) || mag_3d_forced_always);
+
+			if (!_mag_decl_cov_reset) {
+				// After any magnetic field covariance reset event the earth field state
+				// covariances need to be corrected to incorporate knowledge of the declination
+				// before fusing magnetometer data to prevent rapid rotation of the earth field
+				// states for the first few observations.
+				fuseDeclination(0.02f);
+				_mag_decl_cov_reset = true;
+				fuseMag(mag_sample.mag, _aid_src_mag, update_all_states);
+
+			} else {
+				// The normal sequence is to fuse the magnetometer data first before fusing
+				// declination angle at a higher uncertainty to allow some learning of
+				// declination angle over time.
+				fuseMag(mag_sample.mag, _aid_src_mag, update_all_states);
+
+				if (_control_status.flags.mag_dec) {
+					fuseDeclination(0.5f);
+				}
+			}
+
+		} else {
+			// otherwise compute magnetometer innovations for estimator_aid_src_mag logging
+
+			//  rotate magnetometer earth field state into body frame
+			const Vector3f mag_I_body = _state.quat_nominal.rotateVectorInverse(_state.mag_I);
+			const Vector3f mag_observation = mag_sample.mag - _state.mag_B;
+			const Vector3f mag_innov = mag_I_body - mag_observation;
+
+			mag_observation.copyTo(_aid_src_mag.observation);
+			mag_innov.copyTo(_aid_src_mag.innovation);
+		}
+
+
+		// mag heading
+		const float obs_var = fmaxf(sq(_params.mag_heading_noise), 1.e-4f);
+
+		// Rotate the measurements into earth frame using the zero yaw angle
+		Dcmf R_to_earth = updateYawInRotMat(0.f, _R_to_earth);
+
+		Vector3f mag = mag_sample.mag;
+
+		// use mag bias if variance good (unless in HEADING only configured)
+		if (_params.mag_fusion_type != MagFuseType::HEADING) {
+			const Vector3f mag_bias_var = getMagBiasVariance();
+
+			if (mag_bias_var.longerThan(0.f) && !getMagBiasVariance().longerThan(0.001f)) {
+				if (_params.mag_fusion_type == MagFuseType::NONE) {
```

Comment:
```suggestion
				if (_params.mag_fusion_type != MagFuseType::NONE) {
```
BTW, we don't even reach that place of the code in `MagFuseType::NONE`

Reply:
Thanks, these were added in phases after the original testing...

Reply:
This just seems wrong, I wonder if things got confused during a merge. I wanted the mag bias skipped entirely if MagFuseType::HEADING, and otherwise the entire control logic is aborted early if MagFuseType::NONE.

I'm going to remove this conditional.